yamui-screensaverd.c

```diff
@@ -0,0 +1,361 @@
+#include <sys/types.h>
+#include <sys/select.h>
+
+#define UNUSED			__attribute__((unused))
+#define NBITS(x)		((((x) - 1) / __BITS_PER_LONG) + 1)
+#define BIT(arr, bit)		((arr[(bit) / __BITS_PER_LONG] >> \
+				 ((bit) % __BITS_PER_LONG)) & 1)
+
+/*#define DEBUG*/
+
+#define DISPLAY_CONTROL		"/sys/class/graphics/fb0/blank"
+#define DEV_INPUT_DIR		"/dev/input"
+#define LOG_NAME		"screensaverd"
+#define EVENT_PREFIX		"event"
+#define MAX_DEVICES		256
+#define DISPLAY_OFF_TIME	25 /* seconds */
+#define EVENTS_BUF_SIZE		512 /* events */
+#define EXIT_SUCCESS		0
+#define EXIT_FAILURE		1
+
+sig_atomic_t volatile running = 1;
+
+/* ------------------------------------------------------------------------ */
+
+static void
+print_common(const char *format, va_list ap)
+{
+	printf("[%s] ", LOG_NAME);
+	vprintf(format, ap);
+}
+
+/* ------------------------------------------------------------------------ */
+
+static void
+infof(const char *format, ...) __attribute__ ((format (printf, 1, 2)));
+
+/* Info printing with printf-style format and application name prefix. */
+static void
+infof(const char *format, ...)
+{
+	va_list ap;
+	int old_errno = errno;
+
+	va_start(ap, format);
+	print_common(format, ap);
+	printf("\n");
+	va_end(ap);
+	errno = old_errno; /* Just in case. */
+}
+
+/* ------------------------------------------------------------------------ */
+
+#ifdef DEBUG
+
+#define debugf	infof
+
+#else /* !DEBUG */
+
+static void
+debugf(const char *format, ...) __attribute__ ((format (printf, 1, 2)))
+				UNUSED;
+
+static void
+debugf(const char *format UNUSED, ...)
+{
+
+}
+
+#endif /* !DEBUG */
+
+/* ------------------------------------------------------------------------ */
+
+static void
+errorf(const char *format, ...) __attribute__ ((format (printf, 1, 2)));
+
+/* perror() with printf-style format and application name prefix. */
+static void
+errorf(const char *format, ...)
+{
+	va_list ap;
+	int old_errno = errno;
+
+	va_start(ap, format);
+	print_common(format, ap);
+	printf(": %s\n", strerror(old_errno));
+	va_end(ap);
+	errno = old_errno; /* Just in case. */
+}
+
+/* ------------------------------------------------------------------------ */
+
+/* Check for input device type. Returns 0 if button or touchscreen. */
+static int
+check_device_type(int fd)
+{
+	int i;
+	unsigned long bits[EV_MAX][NBITS(KEY_MAX)];
+
+	memset(bits, '\0', sizeof(bits));
+	if (ioctl(fd, EVIOCGBIT(0, EV_MAX), bits[0]) == -1) {
+		errorf("ioctl(, EVIOCGBIT(0, ), ) error on event device");
+		return -1;
+	}
+
+	for (i = 0; i < EV_MAX; i++)
+		if (BIT(bits[0], EV_ABS)) {
+			if (ioctl(fd, EVIOCGBIT(EV_ABS, KEY_MAX),
+				  bits[EV_ABS]) == -1) {
+				errorf("ioctl(, EVIOCGBIT(EV_ABS, ), ) error "
+				       "on event device");
+				return -1;
+			}
+
+			if (BIT(bits[EV_ABS], ABS_MT_POSITION_X) &&
+			    BIT(bits[EV_ABS], ABS_MT_POSITION_Y)) {
+				debugf("Device supports multi-touch events.");
+				return 0;
+			}
+		} else if (BIT(bits[0], EV_KEY)) {
```

Comment:
It is purely theoretical speculation from my part. But since it is possible and dealing with the odd chance is not that difficult in this case, might as well do it.